### PR TITLE
Refs #28643 -- Added Ord, Chr, Left, and Right database functions

### PR DESCRIPTION
**Pull request by bobort**
This is my first time contributing.  I appreciate advice and guidance.

### CODE REVIEW

**Comment by pope1ni** [comment]
I'll endeavour to find some time for this soon as I'm reviewing the other database function submissions.

File: docs/ref/models/database-functions.txt [hunk 1]
```diff
@@ -733,7 +753,24 @@ It can also be registered as a transform. For example::
     >>> # Get authors whose name is longer than 7 characters
     >>> authors = Author.objects.filter(name__length__gt=7)
 
-``Lower``
+
+``Left``
+---------
```
Comment (atombrella) [inline docs/ref/models/database-functions.txt]: Same as above. This is also the reason why the diff has `Lower` in it.

File: docs/releases/2.1.txt [hunk 2]
```diff
@@ -157,6 +157,15 @@ Models
 
 * Models can now use ``__init_subclass__()`` from :pep:`487`.
 
+* Added the :class:~django.db.models.functions.Ord database function.
```
Comment (atombrella) [inline docs/releases/2.1.txt]: You need to add ticks around `~django.db.models.functions.Ord`. Perhaps these four could be merged into a single paragraph?

File: docs/ref/models/database-functions.txt [hunk 3]
```diff
@@ -733,7 +753,24 @@ It can also be registered as a transform. For example::
...
+
+.. class:: Left(expression, length, **extra)
+
+Accepts a single text field or expression and the 1-indexed length of the substring to return.
```
Comment (atombrella) [inline docs/ref/models/database-functions.txt]: I think details about Oracle and MySQL should be mentioned here. Please also wrap text at 79 chars.

File: docs/ref/models/database-functions.txt [hunk 4]
```diff
@@ -785,6 +842,22 @@ Usage example::
     this function) are case-sensitive. Comparisons are case-insensitive by
     default.
 
+``Right``
+---------
+
+.. class:: Right(expression, length, **extra)
+
+Accepts a single text field or expression and the 1-indexed length of the substring to return.
```
Comment (atombrella) [inline docs/ref/models/database-functions.txt]: Same as above with MySQL and Oracle.

File: django/db/models/functions/__init__.py [hunk 5]
```diff
@@ -20,8 +24,11 @@
     'ExtractYear', 'Now', 'Trunc', 'TruncDate', 'TruncDay', 'TruncHour',
     'TruncMinute', 'TruncMonth', 'TruncQuarter', 'TruncSecond', 'TruncTime',
     'TruncYear',
+    # numeric
```
Comment (atombrella) [inline django/db/models/functions/__init__.py]: `Abs` is already proposed in #9577.

File: django/db/models/functions/math.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+from django.db.models import Transform
```
[the remark this answers is not in the document]
Comment (pope1ni) [reply django/db/models/functions/math.py]: Indeed. I suggest that `Round` is moved to a separate PR to land after `Abs`.

File: django/db/models/functions/__init__.py (hunk 5 above)
Comment (bobort) [reply django/db/models/functions/__init__.py]: That was a mistake.  I'll remove it in my next commit.

File: django/db/models/functions/text.py [hunk 7]
```diff
@@ -101,13 +101,78 @@ def __init__(self, expression, pos, length=None, **extra):
...
     lookup_name = 'upper'
+
+
+class Ord(Transform):
+    function = 'ASCII'
+    lookup_name = 'ord'
+    output_field = IntegerField()
+
+    def as_sqlite(self, compiler, connection, **extra_context):
+        return self.as_sql(compiler, connection, function='UNICODE', **extra_context)
```
[the remark this answers is not in the document]
Comment (bobort) [reply django/db/models/functions/text.py]: I don't know how to answer that question.

File: django/db/models/functions/math.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bobort) [reply django/db/models/functions/math.py]: Yes, it was mistake.  I was planning on working on it but decided it against later.

File: django/db/models/functions/text.py [hunk 8]
```diff
@@ -9,10 +9,10 @@ class ConcatPair(Func):
     """
     function = 'CONCAT'
 
-    def as_sqlite(self, compiler, connection):
+    def as_sqlite(self, compiler, connection, **extra_context):
```
[the remark this answers is not in the document]
Comment (timgraham) [reply django/db/models/functions/text.py]: As far as I see, this change is unrelated to this PR, so please remove it.

File: tests/db_functions/tests.py [hunk 9]
```diff
@@ -671,3 +671,159 @@ def test_function_as_filter(self):
             Author.objects.exclude(alias=Upper(V('smithj'))),
             ['Rhonda'], lambda x: x.name
         )
+
+    def test_ord(self):
```
Comment (timgraham) [inline tests/db_functions/tests.py]: Please put these tests in a separate file for each function (e.g. as done for `test_cast.py`).

File: tests/db_functions/tests.py [hunk 10]
```diff
@@ -671,3 +671,159 @@ def test_function_as_filter(self):
...
+
+        self.assertQuerysetEqual(
+            authors.order_by('name'), [
+                'smithj',
+                'rh',
+            ],
+            lambda a: a.alias
+        )
+
+        with self.assertRaisesMessage(ValueError, "'length' must be greater than 0"):
```
Comment (timgraham) [inline tests/db_functions/tests.py]: This should be in a separate test method.

File: tests/test_runner_apps/sample/doctests.py [hunk 11]
```diff
@@ -1,46 +1,46 @@
-"""
```
Comment (timgraham) [inline tests/test_runner_apps/sample/doctests.py]: Please revert these unrelated changes.

File: docs/releases/2.1.txt [hunk 12]
```diff
@@ -172,6 +172,23 @@ Models
 * The new :class:`~django.db.models.functions.Replace` database function
   replaces strings in an expression.
 
+* The new :class:`~django.db.models.functions.Chr` database function returns
+  the unicode character of an integer value.
+
+* The new :class:`~django.db.models.functions.Ord` database function returns
+  the unicode code point of the first character of a string expression.
+
+* Added the :class:`~django.db.models.functions.Left` database function.  By
```
Comment (timgraham) [inline docs/releases/2.1.txt]: Follow the style of the previous notes. "The new ..." Further details like "Oracle and SQLite do not have a Left function.  Django will use their Substr function in its place." should be in the function's documentation rather than the release notes.

File: docs/ref/models/database-functions.txt [hunk 13]
```diff
@@ -762,6 +825,74 @@ Replaces all occurrences of ``text`` with ``replacement`` in ``expression``.
...
+    >>> author = Author.objects.annotate(last_letter=Right('name', 1)).get()
+    >>> print(author.last_letter)
+    h
+
+..
+    Oracle and SQLite do not have a Right builtin function.  Django will
+    substitute Right with the corresponding Substr function to get the
+    same result.
+
+``Replace``
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: This is inadvertently duplicated.

File: docs/ref/models/database-functions.txt [hunk 14]
```diff
@@ -733,6 +780,22 @@ It can also be registered as a transform. For example::
     >>> # Get authors whose name is longer than 7 characters
     >>> authors = Author.objects.filter(name__length__gt=7)
 
+``Left``
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: This is inadvertently duplicated.

File: tests/db_functions/test_chr.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+class ChrTests(TestCase):
+    @classmethod
+    def setUpTestData(cls):
+        Author.objects.create(name='John Smith', alias='smithj')
+        Author.objects.create(name='Élena Jordan', alias='elena')
+        Author.objects.create(name='Rhonda')
+
+    def test_chr(self):
+        authors = Author.objects.annotate(first_initial=Left('name', 1))
+
```
Comment (timgraham) [inline tests/db_functions/test_chr.py]: You can remove many of the blank lines in the tests. The tests are short enough that they're not adding any readability.

File: docs/ref/models/database-functions.txt [hunk 16]
```diff
@@ -762,6 +825,74 @@ Replaces all occurrences of ``text`` with ``replacement`` in ``expression``.
...
+Usage example::
+
+    >>> from django.db.models.functions import Right
+    >>> Author.objects.create(name='Margaret Smith')
+    >>> author = Author.objects.annotate(last_letter=Right('name', 1)).get()
+    >>> print(author.last_letter)
+    h
+
+..
+    Oracle and SQLite do not have a Right builtin function.  Django will
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: I'm not sure if the implementation detail is important for an end user to care about?

File: docs/ref/models/database-functions.txt [hunk 17]
```diff
@@ -675,6 +675,29 @@ that deal with time-parts can be used with ``TimeField``::
 Text functions
 ==============
 
+
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: Chop extra blank line.

File: docs/ref/models/database-functions.txt [hunk 18]
```diff
@@ -675,6 +675,29 @@ that deal with time-parts can be used with ``TimeField``::
...
+``Chr``
+-------
+
+.. class:: Chr(expression, **extra)
+
+.. versionadded:: 2.1
+
+Accepts a single numeric field or expression and returns the text representation
+of the expression as a single character.  It acts equivalently to Python's
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: Please use a single space between all sentences.

File: docs/ref/models/database-functions.txt (hunk 16 above)
Comment (bobort) [reply docs/ref/models/database-functions.txt]: I tend to agree, @timgraham .  @atombrella suggested adding extra information about Oracle and SQLite.  Are you referring to the note about Oracle and SQLite?

File: tests/db_functions/test_chr.py (hunk 15 above)
Comment (bobort) [reply tests/db_functions/test_chr.py]: @timgraham, I have minimized the blank lines as much as possible as PEP 8 recommends. PyCharm also warns me when I remove extra blank lines in violation of PEP 8.  My blank lines in the tests seem to match the spacing of other simple tests as well.  Please let me know how you would like me to proceed.

File: docs/ref/models/database-functions.txt [hunk 19]
```diff
@@ -716,6 +738,25 @@ Usage example::
...
 
+``Left``
+--------
+
+.. class:: Left(expression, length, **extra)
+
+.. versionadded:: 2.1
+
+Accepts a single text field or expression and the 1-indexed length of the
+substring to return. The first ``length`` characters are returned. If length
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: I'm making a few small edits to this PR and don't understand "If length is negative, ``None`` is returned." Is that leftover from an earlier implementation -- now that length is validated to be greater than zero? Or have I missed the meaning?

Comment (bobort) [reply docs/ref/models/database-functions.txt]: Oh, oops.  Yes, that was in a previous commit.

**Comment by timgraham** [comment]
I pushed my edits, and I think this is almost ready. One issue I discovered is that `Chr` doesn't work for non-ascii characters on MySQL and Oracle (see the new failing test). Is there a different database function that should be used there?

File: docs/releases/2.1.txt [hunk 20]
```diff
@@ -194,6 +198,17 @@ Models
 
 * :meth:`.QuerySet.order_by` and :meth:`distinct(*fields) <.QuerySet.distinct>`
   now support using field transforms.
+* The new :class:`~django.db.models.functions.Chr` database function returns
```
Comment (atombrella) [inline docs/releases/2.1.txt]: This is listed twice, both here, and above.

**Comment by felixxm** [comment]
I couldn't reproduce locally test failure on Oracle, IMO everything works properly.

On MySQL `Chr(ord('É'))` gives `CHAR(201 USING utf8)`. If we remove the encoding (`USING utf8`) or we use `utf16` (`USING utf16`) instead of `utf8`, then tests will passed.
